src/EditorFeatures/CSharp/AutomaticCompletion/AutomaticLineEnderCommandHandler.cs

```diff
@@ -273,18 +335,917 @@ private static IEnumerable<SyntaxNode> GetOwningNodes(SyntaxNode root, int posit
+            if (node is DoStatementSyntax { DoKeyword: { IsMissing: false }, Statement: not BlockSyntax } doStatementNode
+                && doStatementNode.DoKeyword.Span.Contains(caretPosition))
+            {
+                return true;
+            }
+
+            if (node is CommonForEachStatementSyntax { Statement: not BlockSyntax, OpenParenToken: { IsMissing: false }, CloseParenToken: { IsMissing: false } }
+                && !WithinEmbeddedStatement(node, caretPosition))
+            {
+                return true;
+            }
+
+            if (node is ForStatementSyntax { Statement: not BlockSyntax, OpenParenToken: { IsMissing: false }, CloseParenToken: { IsMissing: false } }
+                && !WithinEmbeddedStatement(node, caretPosition))
+            {
+                return true;
+            }
+
+            if (node is IfStatementSyntax { Statement: not BlockSyntax, OpenParenToken: { IsMissing: false }, CloseParenToken: { IsMissing: false } }
+                && !WithinEmbeddedStatement(node, caretPosition))
+            {
+                return true;
+            }
+
+            if (node is ElseClauseSyntax elseClauseNode)
+            {
+                // In case it is an else-if clause, if the statement is IfStatement, use its insertion statement
+                // otherwise, use the end of the else keyword
+                // Example:
+                // Before: if (a)
+                //         {
+                //         } else i$$f (b)
+                // After: if (a)
+                //        {
+                //        } else if (b)
+                //        {
+                //            $$
+                //        }
+                if (elseClauseNode.Statement is IfStatementSyntax)
+                {
+                    return ShouldAddBraces(elseClauseNode.Statement, caretPosition);
+                }
+                else
+                {
+                    return elseClauseNode.Statement is not BlockSyntax && !WithinEmbeddedStatement(node, caretPosition);
+                }
+            }
+
+            if (node is LockStatementSyntax { Statement: not BlockSyntax, OpenParenToken: { IsMissing: false }, CloseParenToken: { IsMissing: false } }
+                && !WithinEmbeddedStatement(node, caretPosition))
+            {
+                return true;
+            }
+
+            if (node is UsingStatementSyntax { Statement: not BlockSyntax, OpenParenToken: { IsMissing: false }, CloseParenToken: { IsMissing: false } }
+                && !WithinEmbeddedStatement(node, caretPosition))
+            {
+                return true;
+            }
+
+            if (node is WhileStatementSyntax { Statement: not BlockSyntax, OpenParenToken: { IsMissing: false }, CloseParenToken: { IsMissing: false } }
+                && !WithinEmbeddedStatement(node, caretPosition))
+            {
+                return true;
+            }
+
+            return false;
```

Comment:
as above, i think this should dispatch to all the types, then decide what to do with each type.  this is actually a bit worse as your linear fallthrough means we tests nodes N times even if if could never match all but one case.